### PR TITLE
feat(NODE-4267): support nested fields in type completion for UpdateFilter

### PR DESCRIPTION
**Pull request by coyotte508**
#### What is changing?

The `MatchKeysAndValues` type used by `UpdateFilter` is changed, inspired by the `Filter` type for querying.

##### Is there new documentation needed for these changes?

No

#### What is the motivation for this change?

The `Filter` type for querying is very nice. But the `UpdateFilter` type doesn't support type completion for nested fields. 

https://jira.mongodb.org/projects/NODE/issues/NODE-4267

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/mongo_types.ts [hunk 1]
```diff
@@ -258,7 +258,26 @@ export type OnlyFieldsOfType<TSchema, FieldType = any, AssignableType = FieldTyp
 >;
 
 /** @public */
-export type MatchKeysAndValues<TSchema> = Readonly<Partial<TSchema>> & Record<string, any>;
+export type MatchKeysAndValues<TSchema> = Readonly<{
+  [Property in Join<NestedPaths<TSchema>, '.'>]?: PropertyType<TSchema, Property>;
+}> &
```
Comment (coyotte508) [inline src/mongo_types.ts:263]: The `ReadOnly<...>` could be extracted, eg:

```
type MatchKeysAndValuesNoPositional<TSchema> =  Readonly<{
  [Property in Join<NestedPaths<TSchema>, '.'>]?: PropertyType<TSchema, Property>;
}>

type MatchKeysAndValues<TSchema> = MatchKeysAndValuesNoPositional<TSchema> & Partial<
    Record<
      `${KeysOfAType<MatchKeysAndValuesNoPositional<TSchema>, any[]>}.$${`[${string}]` | ''}${
        // Subsequent nested path - could be narrowed
        `.${string}` | ''
      }`,
      // Value for updated field - could be narrowed
      any
    >
  >;
```

But that would make an additional exported type

File: src/mongo_types.ts [hunk 2]
```diff
@@ -258,7 +261,19 @@ export type OnlyFieldsOfType<TSchema, FieldType = any, AssignableType = FieldTyp
...
+  {
+    [Property in Join<NestedPaths<TSchema>, '.'>]?: PropertyType<TSchema, Property>;
+  } & {
+    [Property in `${NestedPathsOfType<TSchema, any[]>}.$${`[${string}]` | ''}`]?: ArrayElement<
+      PropertyType<TSchema, Property extends `${infer Key}.$${string}` ? Key : never>
+    >;
+  } & {
+    [Property in `${NestedPathsOfType<TSchema, Record<string, any>[]>}.$${
+      | `[${string}]`
+      | ''}.${string}`]?: any; // Could be further narroyed
```
Comment (julien-c) [inline src/mongo_types.ts:274]:
```suggestion
      | ''}.${string}`]?: any; // Could be further narrowed
```

**Comment by baileympearson** [comment]
Hey @julien-c!  Thanks for this contribution.  The Node team is interested in adding this feature to the driver but we don't have the bandwidth at the moment to evaluate the changes and get it merged.  

We plan on scheduling this work - you can track the status in the Jira ticket you filed (https://jira.mongodb.org/browse/NODE-4267).

**Comment by coyotte508** [comment]
@baileympearson I updated the branch - feel free to update it too :)

**Comment by baileympearson** [comment]
Thanks - and sorry for pinging the wrong user about this PR!

**Review by baileympearson — APPROVED** [review]

**Comment by baileympearson** [comment]
@coyotte508 Thanks for taking the time to help us with this and thank you for being patient with us!

**Comment by coyotte508** [comment]
There's a regression when wanting to update nested objects directly:

```ts
interface User {
  subscription: {
    paymentMethod: {
      last4: string;
      expiration: Date;
    }
  }
}
```

I can update `subscription.paymentMethod.last4` or `subscription.paymentMethod.expiration` invidiually but not `subscription.paymentMethod`.

**Pull request merged**